facebook/react-native
# Fix ios build error when use_frameworks is on and fabric is off

## Summary

when `use_frameworks!` is on, there are errors like:
```
'FBReactNativeSpec/FBReactNativeSpec.h' file not found
#import <FBReactNativeSpec/FBReactNativeSpec.h>
```
this error may come from from https://github.com/facebook/react-native/commit/f7e4c07c84b6 regression.


when `use_frameworks!` is on, xcode will search headers from framework directories, the correct imports would be `#import <React_Codegen/FBReactNativeSpec/FBReactNativeSpec.h>` (xcode will transform dash to underscore, so it is `React_Codegen` but not `React-Codegen`). in the other hand, when `use_frameworks!` is off, the correct import is `#import <React-Codegen/FBReactNativeSpec/FBReactNativeSpec.h>`.

~to make it compatible for both modes, i would propose to use a valid module name for the codegen pod. that is the `ReactCodegen`. supposedly can define different module name in podspecs either by [`header-dir`](https://guides.cocoapods.org/syntax/podspec.html#header_dir) or [`module_name`](https://guides.cocoapods.org/syntax/podspec.html#module_name). however, i don't recommend to use these as [it has a known issue when DEFINES_MODULE=YES](https://github.com/CocoaPods/CocoaPods/issues/11136). that's why i use the `ReactCodegen` name.~
Update: to make it compatible for both modes, i merged `FBReactNativeSpec` and `React-Codegen`. the final name is `FBReactNativeSpec.podspec`, so the final valid import will be `#import <FBReactNativeSpec/FBReactNativeSpec/FBReactNativeSpec.h>`. this is specific for `header_mappings_dir = '.'` mode. to keep the import as original `#import <FBReactNativeSpec/FBReactNativeSpec.h>`, i removed `header_mappings_dir` in non-fabric mode. the workaround has a downside where assuming there're no other codegen libraries. otherwise, it may also break imports, e.g. `#import <ScreenshotmanagerSpec/ScreenshotmanagerSpec.h> -> #import <FBReactNativeSpec/ScreenshotmanagerSpec.h>`.

this fix is specific for old architecture (fabric is off).

when fabric is on, there are other errors from duplicated headers when copying to build folder. [the reason is that framework build would try to flatten headers](https://mkonrad.net/2015/03/29/xcode-static-libraries-preserving-header-directory-structure.html). we have `primitives.h` in different folders and they would be flattened into `React_Fabric.framework/Headers`. to be honest, i don't know how to deal with the problem in the meantime,  maybe subspecs are not enough, we should separate them from subspecs to dedicated podspecs so that we can have these targets as different frameworks.

## Changelog

[iOS] [Fixed] - Fix iOS build error when Podfile `use_frameworks!` is on and Fabric is off

## Test Plan

verify with rn-tester
1. change `fabric_enabled` to false in `packages/rn-tester/Podfile`
2. `USE_FRAMEWORKS=1 pod install`
3. build rn-tester in xcode
